libkbfs/chat_rpc.go

```diff
@@ -0,0 +1,298 @@
+			KeybaseServiceName, errDisconnected{})
+	}
+}
+
+// ShouldRetry implements the ConnectionHandler interface.
+func (c *ChatRPC) ShouldRetry(_ string, _ error) bool {
+	return false
+}
+
+// ShouldRetryOnConnect implements the ConnectionHandler interface.
+func (c *ChatRPC) ShouldRetryOnConnect(err error) bool {
+	_, inputCanceled := err.(libkb.InputCanceledError)
+	return !inputCanceled
+}
+
+var _ Chat = (*ChatRPC)(nil)
+
+// Chat notes (will remove/rework these once the implementation is complete):
+//
+// When sending:
+//   * chat1.NewConversationLocal
+//   * chat1.PostTestNonblock
+//     * ClientPrev can be 0.  Can outbox ID be nil?
+
+// Gathering recent notifications:
+//   * chat1.GetInboxAndUnboxLocal (pagination not needed)
+//     * But we'd need inbox grouping to get this exactly right.
+
+// Reading each conversation:
+//   * Get list of all channels/writers in the conversation
+//     * chat1.GetTLFConversationsLocal can give us the list of channels,
+//       which corresponds to the set of users who have actually written.
+//       (There might be a lot of team members who haven't written, so
+//       probably best to avoid iterating through everyone.).
+//     * Always look up your own channel though, so the GUI can show your
+//       own last edits if desired.
+//   * Read each channel until getting N updates for each writer
+//     * chat1.GetThreadLocal with pagination
+//     * No prev filled in on next pagination to go backward
+//   * How to reconcile renames, etc across channels?
+//     * It's hard to know if a long-ago writer updated a file, and
+//       later it was renamed by a prolific writer who made N more updates
+//       afterward.
+//     * For performance reasons, we probably just need to show the old
+//       update under the old file name.  Should be rare enough that
+//       it doesn't matter.
+
+// Getting real-time updates:
+//   * New kbfs-edits activity push notifications on notify-router
+//   * Mike will make ticket to auto-join non-chat channels,
+//     so they show up in `GetInboxAndUnboxLocal`.
+//   * Spot-edit the local edit history on each new push notification.
+
+// One layer over the service RPC connection that takes all needed
+// arguments (including topic type, etc), and passes it pretty
+// directly to the RPC.
+
+// Another, per-TLF layer to remember the resolved conversation ID and
+// send/read kbfs-edits messages.  It should also interface with the
+// local journal to show the unflushed journal data as part of the
+// updates.
+
+// Finally an inbox layer that can read the server inbox, and also
+// checks the journal status, to return the top set of conversations
+// at any given time.  Maybe it also subscribes to inbox notifications
+// of some kind.
+
+func membersTypeFromTlfType(tlfType tlf.Type) chat1.ConversationMembersType {
+	if tlfType == tlf.SingleTeam {
+		return chat1.ConversationMembersType_TEAM
+	}
+	return chat1.ConversationMembersType_IMPTEAMNATIVE
```

Comment:
what about `chat1.ConversationMembersType_IMPTEAMUPGRADE`?

Reply:
This behavior is what Mike told me to do.  I think `IMPTEAMUPGRADE` is a temporary core-internal thing, if I understand it correctly, just used during migration.

Reply:
OK cool, thanks.

Reply:
I think there's a difference between the teams that are upgraded from an old TLF (`impteamupgrade`) and teams that are born so (`impteamnative`). IIRC they'll stay like that. For example, my chat has 103 `impteamupgrade` conversations.

Reply:
Ah ok thanks.  But in this case, since it will always be new notifications, we never need to mark them as upgraded.

Reply:
You mean notifications use conversations that are newly created from now, and they'll always be `impteamnative` since it's enabled globally now? Yeah that sounds good.

Reply:
OK now I'm confused. I thought we're using existing conversations, just different `topic-type`s.

Reply:
I'm pretty sure a new topic type implies a new underlying conversation (though it could of course have the same name as some existing conversation).

Reply:
Analogy: private TLFs vs public TLFs.

Reply:
I thought new conversation implied new key set. Topic types are just channels that are separate from the crypto aspect. Maybe @mmaxim could elucidate.

Reply:
Either way, the thing that matters is that it's a new conversation ID (every channel has a separate ID as well), and the members type is specific to the ID I believe.

Reply:
Also, a "conversation ID" is equal to `SHA256((TeamID, Topic Type, Topic ID))`